hail/src/main/scala/is/hail/expr/ir/InferPType.scala

```diff
@@ -81,7 +158,7 @@ object InferPType {
           val required = l.pType2.required && r.pType2.required
           val vType = BinaryOp.getReturnType(op, l.pType2.virtualType, r.pType2.virtualType).setRequired(required)
 
-          PType.canonical(vType)
+          PType.canonical(vType, vType.required)
```

Comment:
Why doesn't this propagate the ptype of the body?

Reply:
What is the body, the operation? It's propagating the requiredeness of the operands, you're saying it should rely on the requiredeness of the operation as well? 

The operation ( or condition in the If node case) just specifies the kind of operation to be performed. I think it's requiredeness should be trivially set, meaning it doesn't really have requiredeness. If you specify it, that operation should always be performed, if possible, and "if possible" is a property of node that operation is being applied on (ApplyBinaryPrimOp). 

So in this case, if both operands required, then ApplyBinaryPrimOp is required. If I understand what you're saying, for consistency I could set `op.required == l.pType2.required && r.pType2.required`, then set ApplyBinaryPrimOp's requiredeness to == `op.required`. Or I could choose to always make `op` required, or always optional, it wouldn't matter because requiredeness doesn't seem to mean anything for the operation in isolation.

Could you help me understand your perspective?

Reply:
Haha, I got confused here by Github collapsing unchanged lines. I thought this was the return type for the `Let` node, my bad!

Reply:
I'm guessing Arcturus thought this was still the `Let` case. But note that requiredness of `op` doesn't make sense. `op` is not an `IR`, it's a `BinaryOp`. In particular `op` is a compile-time value. This node only has two run-time inputs which may or may not be missing.

More formally, requiredness is a property of hail types. `l` and `r` have hail types (as does everything of scala type `IR`), but `op: BinaryOp` does not.

Reply:
Yep, this is true, I checked and realized this after I wrote the comment, thanks.